### ISSUE TITLE
Export disjoint valid/missing/anomalous dataframes 

### ISSUE THREAD

**Issue by GjjvdBurg** [alan-turing-institute/ptype-dmkd#11]
Projection:
- [x] for each projection, map discarded values to `pd.NA`
- [x] first row of `auto.csv` is being treated as column header
  - [x] convert column “numbers” to strings
- [x] move `as_missing` etc from test code to `Ptype`

Design/implementation improvements, incl. steps towards a less imperative API:
- [x] delete `get_categorical_signal_gaussian` (unused)
- [x] merge `change_missing_data_annotations` and `change_anomaly_annotations`

### Done/dropped

Misc
- [x] informal specification of how we split the dataframe into 3 (see picture below)
- [x] move notebook tests inside `test_ptype.py`

Initial experiments to grok current design
- [x] ensure easy to run tests inside PyCharm
- [x] after `runInference` (and thus `storeOutputs`), inspect contents of `self.[missing|anomaly]_types`
  - [x] for each `column_name` in `enumerate(list(self.model.experiment_config.column_names)`
  - [x] union of (normal, missing, anomaly) should be original set of row indices :x:
  - for each (column, row) in original dataframe:
    - [x] show value
    - [x] find index in “unique values” for that column
      - [x] `KeyError` with `OrgName` column
        - [x] delete occurrences of `data_frame.rename(columns=lambda n: str(n).replace(‘ ‘, ‘’)`
      - [x] check `get_unique_vals(col)` and `missing_types[col]` are both lists
      - [x] restrict `get_missing_data_predictions` and `get_anomaly_predications` to a single column
      - [x] `missing_types[col]` must be a subset of `range(get_unique_vals(col))`
    - [x] show whether (normal, missing, anomaly)

Design/implementation improvements, incl. steps towards a less imperative API:
- [x] per-row classification ∈ {normal, missing, anomaly} more robust than three-way partition of row indices
- [x] inline `change_column_type_annotations` (simply sets `predicted_type`)
- [x] `get_columns_with_missing`: express as loops over `Column`
- [x] delete old `Ptype.show_results` and just demo notebook accordingly
- [x] `initialize_params_randomly` no longer required
- [x] cache `unique_vals` and `unique_vals_counts` on `Column`
- [x] delete `set_data_multiple_dfs` (unused)? 
- [x] factor out `np.unique([str(int_element) for int_element in self.model.data[col].tolist()])` (unique vals)
- [x] remove redundant initialisation of `anomaly_types`, `predicted_types` etc inside `Ptype`
- [x] remove redundant parameters from `set_data`
- [x] drop `data_frames` parameter from `Ptype_init`
- [x] delete `get_empty_columns`, `get_empty_columns_with anomalies`, etc
- [x] remove redundancy in `show_results`
- [x] delete `run_experiment`, `read_dataset`, `setup_a_column` (unused)
- [x] consolidate `set_data`, `__init__` and `run_inference`
- [x] purge `_print`, `_predictionPath` parameters
- [x] `Config` doesn’t need default value for `types`
- [x] delete both versions of `train_machine_multiple_dfs`, `train_all_models_multiple_dfs` (unused)
- [x] `write_type_predictions_2_csv` – remove “append” case; consolidate other two cases
- [x] delete unused `self.cats` assignment in `get_categorical_columns`
- [x] simplify `get_empty_columns`, `get_columns_with_anomalies`, etc
- [x] use `add_to_normal`, `remove_from_missing` to self-document `change_[missing|anomaly]_annotations`
- [x] rename `merge_missing_data` to `replace_missing`
- [x] delete `get_empty_columns` (unused)
- [x] delete `get_unnormal_data_indices` (unused)
- [x] delete `p_z_columns` and `p_t_columns` (unused)
- [x] make normalisation methods in `PtypeModel` static
- [x] introduce `ColResult` and migrate column-specific behaviour there

***
Original suggestion by @GjjvdBurg:
> One idea for how users could interact with ptype would be using the scikit-learn transformer model. Transformers have a ``fit`` method that does the learning/fitting of a model and a ``transform`` method that uses that model to transform an input. A simple example is the [MinMaxScaler](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html#sklearn.preprocessing.MinMaxScaler) that scales a matrix. Transformers in scikit-learn inherit the [TransformerMixin](https://scikit-learn.org/stable/modules/generated/sklearn.base.TransformerMixin.html#sklearn.base.TransformerMixin), which provides the combined ``fit_transform`` method.

> I could imagine a ``PType`` class that provides a ``fit`` method to learn the model/weights (if I understood correctly) and a ``transform`` method that uses the MAP estimate to cast an input matrix to a typed array (dataframe?). The ``PType`` object can then hold state, and provide other methods to interact with the ptype model besides fit/transform. If you’d want to work with dataframes, it might be worthwhile to see if the new (experimental) [``pd.NA``](https://pandas.pydata.org/pandas-docs/stable/user_guide/missing_data.html#missing-data-na) type can be helpful.

Also see [Open Source Data Project Developers](https://discuss.ossdata.org/) for a community that might have some useful opinions.

Here’s a picture (adapted from @tahaceritli) that shows the partitioning idea:
![Screenshot 2020-05-15 at 09 49 41](https://user-images.githubusercontent.com/121074/82031324-a1f78900-9691-11ea-8597-67204da831df.png)

**Comment by rolyp** [comment]
Adding note from @triangle-man's email. (Suggest any further discussion takes place here, or on relevant issues as we create them.)

> ptype returns columns whose type is not supported by R or pandas.
> 
> I was thinking about this, especially in the context of “how do we make it easy to adopt these tools” and “how do Wrattler assistants talk to each other?”
> 
> Below are some random thoughts.
> 
> 
> # Datasets
> 
> It strikes me that the thing we are all used to thinking about as a “dataset” –
> the thing with rows and columns – actually comes in at least three guises, none
> of which are what we _actually_ want.
> 
> ## Three similar models of data
> 
> The “statistician’s model” of data is this: a dataset is a set of “cases”. Each
> cases is (or possibly is characterised by) a tuple of observations. The tuple of
> observations must describe the same set of observations for all cases. It is
> typically unclear whether the tuple of values associated with a particular case
> _define_ that case (that is, whether they individuate it), or whether the
> existence of a row indicates the existence of a unique entity in the world that
> just _happens_ to have the characteristics given by the observations.
> 
> The “data scientist’s model” is: a dataset is a set of columns, all of the same
> length. Each column is a sequence of values (or “vector” in the jargon), all of
> the same type. The columns represent particular kinds of observations. Given an
> index, $i$, the set of values of the columns at index $i$ represent the set of
> observations on some “case.”
> 
> By _type_ here is mean one of the primitive, atomic types supported by the computer
> system, such as “bounded integer,” “IEEE 754 floating-point number,” and so
> on. Sometimes the values are allowed to be a special value in addition to the
> base type, indicating a missing.
> 
> The “relational database model” is that a dataset is a _relation_: that is, it
> is a subset of the Cartesian product of some sets. The sets in the Cartesian
> product represent the basic types. This model is most like the statistician’s
> model, except that the types are allowed to be arbitrary sets.
> 
> ## What’s missing?
> 
> Two things are missing: (1) Some columns are not atomic types; (2) Some types
> cannot be represented as columns.
> 
> The first is the thing that ptype tells us. Some columns are not just “floats,”
> or even “floats plus a missing”, they are “floats, or a missing, or an anomalous
> value, or some different kind of missing.” That is, some types are _union types_
> (aka _sum types_).
> 
> The second is similar. Some kinds of value are represented as _multiple_
> columns. Coordinates are a good example. It’s weird to have, say, latitude and
> longitude treated by the computer as completely independent entities. Really,
> they are two components of the same value. In other words, some types are
> _record types_ (aka _product types_).
> 
> 
> # Proposal
> 
> ## ptype and union types
> 
> What I think is important for ptype is a way to represent, in the “data
> scientist’s model,” how ptype itself thinks of the world. That is, we need a way
> for ptype to return a table whose columns are union types, when the system does
> not support union types.
> 
> There seem to be two possible approaches, neither great.
> 
> First, suppose only one column, column A, were affected, and that column A
> consists of values of type A_1 and other values of type A_2. Then we could split
> the table into two tables, having identical column names, where the affected
> column is of type A_1 in the first table and type A_2 in the second table. In
> effect, we are saying that the “real” table is the union of these two tables,
> which is true. Even with types, $X\times(A_1+A_2) = X\times A_1 + X\times A_2$,
> 
> If there were two affected columns, A and B, where again the type of each column
> were the union of two types, then we could split the table into four tables:
> $X\times A_1 \times B_1$, $X\times A_1 \times B_2$, and so on.
> 
> In general, if the type of column $X$ is the union of $\Pi_X$ types, then one
> ends up with $\Pi_A \times Pi_B \times \dotsc$ separate tables. That can be
> quite a lot of tables.
> 
> Alternatively, one could break each _column_ into multiple columns, each one
> hold either a value of that type, or whatever value codes for “missing”. (In R,
> any primitive type is allowed to be `NA`.) If the language doesn’t support
> missings, then one could add an indicator column saying which column has the
> actual value.
> 
> This results in a single table with $\Pi_A + \Pi_B + \cdots$ columns.
> 
> The first approach makes it hard to get all the values of a single type from a
> single column, because those values might have been split by the types in a
> different column and be in different tables. On the other hand, one can treat
> each table as a set of observations of a single type; I can imagine that makes
> it easier to do Bayesian modelling, for example.
> 
> The second approach doesn’t have so much proliferation of entities.
> 
> ### What’s to be done
> 
> - Theory
> - Naming convention for new columns
> - Metadata for capturing the structure
> - Utility functions for managing such tables
> - Extensions to tidyverse/pandas?
> 
> ## Record types
> 
> Haven’t thought about this one yet!

**Comment by rolyp** [comment]
And adding a reply to @triangle-man from @tpetricek!

> I was also wondering, some time ago, about what ptype could do to handle unions and records (essentially, the two things that would be needed if we wanted to extend ptype to work with JSON and XML files in the same way in which F# Data does that).
> 
> The tricky thing here is that the way ptype works (AFAIK) now is that it has a list of possible types (which are constructed as core type | outlier | missing). It then iterates over those types to see what best matches a given list of values.
> 
> This no longer works with records and unions, because those are composite types - composed from other more primitive types - and so the space of possible types is just too big (or possibly infinite). This means that you cannot meaningfully iterate over all types.
> 
> I’m not sure how this could be handled - the core PFSM model is elegant, but it does not have a good way of dealing with this. Maybe there would be some way to split the inference process into two steps, one that would ((somehow) probabilistically) split the input data (document) into multiple sub-documents on which the individual PFSMs can then be run to get the individual types and then recompose these? I don’t really know - but I think this issue of composite types is an important problem!
>

**Comment by rolyp** [comment]
@GjjvdBurg @tahaceritli @triangle-man Renamed this task now that we have a clearer idea of the goal.
